File: src/shared/types/storage.ts
export type FileInfo = {
  name: string;
  path: string;
  isDirectory: boolean;
  size: number;
  modifiedAt: number;
};

export type AssetDiskFileInfo = {
  projectName: string;
  categoryName: string;
  name: string;
  relativePath: string;
  size: number;
  modifiedAt: number;
};

export type AssetDiskProjectInfo = {
  name: string;
  relativePath: string;
  createdAt: number;
  modifiedAt: number;
};

export type StorageResult<T = unknown> = {
  success: boolean;
  error?: string;
} & T;

export type StoryboardAssetsPackageResult = StorageResult<{
  assets?: unknown;
  path?: string;
  canceled?: boolean;
  copiedMediaCount?: number;
}>;

export type StorageApi = {
  selectDirectory: () => Promise<string | null>;
  ensureProject: (
    basePath: string,
    projectName: string,
  ) => Promise<StorageResult<{ project?: unknown; path?: string }>>;
  listProjects: (basePath: string) => Promise<
    StorageResult<{
      projects?: Array<{ name: string; createdAt: number; updatedAt: number }>;
    }>
  >;
  saveCanvas: (
    basePath: string,
    projectName: string,
    data: unknown,
  ) => Promise<StorageResult>;
  loadCanvas: (
    basePath: string,
    projectName: string,
  ) => Promise<StorageResult<{ data: unknown | null }>>;
  saveMedia: (
    basePath: string,
    relativePath: string,
    buffer: ArrayBuffer,
  ) => Promise<StorageResult<{ path?: string }>>;
  readMedia: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult<{ data: Buffer | null }>>;
  writeRawFile: (
    basePath: string,
    relativePath: string,
    buffer: ArrayBuffer,
  ) => Promise<StorageResult<{ path?: string }>>;
  readRawFile: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult<{ data: Buffer | null }>>;
  scanAssetLibrary: (
    basePath: string,
  ) => Promise<
    StorageResult<{
      projects?: AssetDiskProjectInfo[];
      files?: AssetDiskFileInfo[];
    }>
  >;
  deleteRawPath?: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult>;
  renameRawPath?: (
    basePath: string,
    oldRelativePath: string,
    newRelativePath: string,
  ) => Promise<StorageResult>;
  listMedia: (
    basePath: string,
    projectName: string,
    mediaType: string,
  ) => Promise<StorageResult<{ files?: FileInfo[] }>>;
  deleteMedia: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult>;
  downloadMedia: (
    basePath: string,
    url: string,
    relativePath: string,
  ) => Promise<StorageResult<{ path?: string }>>;
  saveBufferToFile: (
    defaultFileName: string,
    buffer: ArrayBuffer,
  ) => Promise<StorageResult<{ path?: string; canceled?: boolean }>>;
  mediaExists: (basePath: string, relativePath: string) => Promise<boolean>;
  renameProject: (
    basePath: string,
    oldProjectName: string,
    newProjectName: string,
  ) => Promise<StorageResult>;
  deleteProject: (
    basePath: string,
    projectName: string,
  ) => Promise<StorageResult>;
  copyProject: (
    basePath: string,
    srcProjectName: string,
    destProjectName: string,
  ) => Promise<StorageResult>;
  exportProject: (
    basePath: string,
    projectName: string,
  ) => Promise<
    StorageResult<{ path?: string; projectName?: string; canceled?: boolean }>
  >;
  importProject: (
    basePath: string,
  ) => Promise<
    StorageResult<{ path?: string; projectName?: string; canceled?: boolean }>
  >;
  exportStoryboardAssets: (
    basePath: string,
    projectId: string,
    assets: unknown,
  ) => Promise<StoryboardAssetsPackageResult>;
  importStoryboardAssetsPackage: (
    basePath: string,
    projectId: string,
  ) => Promise<StoryboardAssetsPackageResult>;
  getDefaultPath: () => Promise<string>;

  // Legacy aliases retained for compatibility while the renderer migrates.
  ensureProjectDir: (
    basePath: string,
    projectName: string,
  ) => Promise<StorageResult<{ project?: unknown; path?: string }>>;
  writeFile: (
    basePath: string,
    relativePath: string,
    buffer: ArrayBuffer,
  ) => Promise<StorageResult<{ path?: string }>>;
  readFile: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult<{ data: Buffer | null }>>;
  deleteFile: (
    basePath: string,
    relativePath: string,
  ) => Promise<StorageResult>;
  listFiles: (
    basePath: string,
    projectName: string,
    mediaType: string,
  ) => Promise<StorageResult<{ files?: FileInfo[] }>>;
  downloadFile: (
    basePath: string,
    url: string,
    relativePath: string,
  ) => Promise<StorageResult<{ path?: string }>>;
  fileExists: (basePath: string, relativePath: string) => Promise<boolean>;
};

File: src-tauri/src/models/storage.rs
use serde::{Deserialize, Serialize};

/// 与前端 StorageApi 完全对等的领域模型。

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProjectMeta {
    pub name: String,
    #[serde(rename = "createdAt")]
    pub created_at: u64,
    #[serde(rename = "updatedAt")]
    pub updated_at: u64,
    #[serde(rename = "coverUrl", skip_serializing_if = "Option::is_none", default)]
    pub cover_url: Option<String>,
    #[serde(rename = "coverLocalPath", skip_serializing_if = "Option::is_none", default)]
    pub cover_local_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProjectIndex {
    pub version: u32,
    pub projects: std::collections::BTreeMap<String, ProjectMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileInfo {
    pub name: String,
    pub path: String,
    #[serde(rename = "relativePath")]
    pub relative_path: String,
    #[serde(rename = "mediaType")]
    pub media_type: String,
    #[serde(rename = "isDirectory")]
    pub is_directory: bool,
    pub size: u64,
    #[serde(rename = "modifiedAt")]
    pub modified_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetDiskFileInfo {
    #[serde(rename = "projectName")]
    pub project_name: String,
    #[serde(rename = "categoryName")]
    pub category_name: String,
    pub name: String,
    #[serde(rename = "relativePath")]
    pub relative_path: String,
    pub size: u64,
    #[serde(rename = "modifiedAt")]
    pub modified_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetDiskProjectInfo {
    pub name: String,
    #[serde(rename = "relativePath")]
    pub relative_path: String,
    #[serde(rename = "createdAt")]
    pub created_at: u64,
    #[serde(rename = "modifiedAt")]
    pub modified_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct StorageResult<T = ()> {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(flatten)]
    pub data: T,
}

impl<T: Default> StorageResult<T> {
    pub fn ok() -> Self {
        Self { success: true, error: None, data: T::default() }
    }

    pub fn err(msg: impl Into<String>) -> Self {
        Self { success: false, error: Some(msg.into()), data: T::default() }
    }
}
